src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -1344,676 +1431,434 @@ private bool IsMiniShell()
-                        }
-                        else
-                        {
-                            string messageData = null;
-                            try
-                            {
-                                messageData = (string)LanguagePrimitives.ConvertTo(obj, typeof(string), CultureInfo.InvariantCulture);
-                            }
-                            catch (PSInvalidCastException)
-                            {
-                                continue;
-                            }
-
-                            obj = new InformationRecord(messageData, null);
-                        }
-                    }
-                    else if (stream == MinishellStream.Debug ||
-                             stream == MinishellStream.Verbose ||
-                             stream == MinishellStream.Warning)
-                    {
-                        //Convert to string
-                        try
-                        {
-                            obj = LanguagePrimitives.ConvertTo(obj, typeof(string), CultureInfo.InvariantCulture);
-                        }
-                        catch (PSInvalidCastException)
-                        {
-                            continue;
-                        }
-                    }
-                    AddObjectToWriter(obj, stream);
+                    _streamWriter.Dispose();
                 }
-            }
-            catch (XmlException originalException)
-            {
-                string template = NativeCP.CliXmlError;
-                string message = string.Format(
-                    null,
-                    template,
-                    _isOutput ? MinishellStream.Output : MinishellStream.Error,
-                    _processPath,
-                    originalException.Message);
-                XmlException newException = new XmlException(
-                    message,
-                    originalException);
-
-                ErrorRecord error = new ErrorRecord(
-                    newException,
-                    "ProcessStreamReader_CliXmlError",
-                    ErrorCategory.SyntaxError,
-                    _processPath);
-                AddObjectToWriter(error, MinishellStream.Error);
-            }
-        }
-
-        /// <summary>
-        /// Adds one object to writer
-        /// </summary>
-        private void AddObjectToWriter(object data, MinishellStream stream)
-        {
-            try
-            {
-                ProcessOutputObject dataObject = new ProcessOutputObject(data, stream);
-                //writer is shared between Error and Output reader.
-                lock (_writer)
+                catch (IOException)
                 {
-                    _writer.Write(dataObject);
+                    // on unix, if process is already finished attempt to dispose it will
+                    // lead to "Broken pipe" exception.
+                    // we are ignoring it here
```

Comment:
Again, speaking of exception perf penalty, how often this will happen on unix platforms? If it's often enough, maybe checking `process.HasExited` before disposing (only for unix) would be preferable?